File: frontend/src/types/analysis.ts
export interface Summary {
  ip: string;
  country: string | null;
  organization: string | null;
  threat_score: number;
  risk_level: string;
  recommendation: string;
}

export interface IPInfo {
  ip: string;

  city: string | null;
  region: string | null;
  country: string | null;

  organization: string | null;
  asn: string | null;
  hostname: string | null;

  latitude: number | null;
  longitude: number | null;
}

export interface AbuseIPDB {
  abuse_confidence_score: number | null;
  total_reports: number | null;
  last_reported_at: string | null;
}

export interface VirusTotal {
  malicious: number | null;
  suspicious: number | null;
  harmless: number | null;
  undetected: number | null;
  reputation: number | null;
}

export interface GreyNoise {
  classification: string | null;
  name: string | null;
}

export interface Shodan {
  ports: number[];
  tags: string[];
}

export interface AnalysisSources {
  ipinfo: IPInfo;
  abuseipdb: AbuseIPDB;
  virustotal: VirusTotal;
  greynoise: GreyNoise;
  shodan: Shodan;
}

export interface AnalysisResponse {
  summary: Summary;
  sources: AnalysisSources;
}

File: backend/app/schemas/analysis.py
from pydantic import BaseModel

from app.schemas.summary import Summary
from app.schemas.ipinfo import IPInfoResponse
from app.schemas.abuseipdb import AbuseIPDBResponse
from app.schemas.virustotal import VirusTotalResponse
from app.schemas.greynoise import GreyNoiseResponse
from app.schemas.shodan import ShodanResponse


class Sources(BaseModel):
    ipinfo: IPInfoResponse
    abuseipdb: AbuseIPDBResponse
    virustotal: VirusTotalResponse
    greynoise: GreyNoiseResponse
    shodan: ShodanResponse


class AnalysisResponse(BaseModel):
    summary: Summary
    sources: Sources